### PR TITLE
Add build/test instructions

### PR DESCRIPTION
**Pull request by say4n**
Added build/test instruction to the main README.md.

Presently only CONTRIBUTING.md [has some instructions and hints](https://github.com/mlpack/ensmallen/blob/master/CONTRIBUTING.md#buildtest-process) at the main README.md containing more information on the same (which it doesn't).

### CODE REVIEW

**Comment by zoq** [comment]
Personally, I would only update the `CONTRIBUTING.md` file, as ensmallen is a header-only library, so besides armadillo there is no other dependency like cmake; I don't think building and running the tests is relevant for someone who just likes to use ensmallen.

`"Information on how to build and run the tests is in the main README.md file."`

Looks like we haven't put that information in the `README.md` file, so maybe we should remove that line and just update the file with the information you put together.

File: README.md [hunk 1]
```diff
@@ -15,6 +15,96 @@ gradient-free optimizers, and constrained optimization.
...
+
+<details open>
+<summary>Linux</summary>
+
+Use your distributions' package manager to install the required dependencies. The commands shown below are for Ubuntu.
+
+```bash
+$ sudo apt-get update
+$ sudo apt-get install -y g++ cmake
+$ sudo apt-get install -y libopenblas-dev liblapack-dev xz-utils
```
Comment (zoq) [inline README.md]: No need to install those packages, armadillo is already packaged with a BLAS variant, and xz utils are only necessary to extract the armadillo archive, which isn't needed if you install it via the package manager.

File: README.md [hunk 2]
```diff
@@ -15,6 +15,96 @@ gradient-free optimizers, and constrained optimization.
...
+### Install Dependencies
+
+<details open>
+<summary>Linux</summary>
+
+Use your distributions' package manager to install the required dependencies. The commands shown below are for Ubuntu.
+
+```bash
+$ sudo apt-get update
+$ sudo apt-get install -y g++ cmake
```
Comment (zoq) [inline README.md]: Not sure we should use `-y` here, a user might want to control the install process.

File: README.md [hunk 3]
```diff
@@ -15,6 +15,96 @@ gradient-free optimizers, and constrained optimization.
...
+```
+</details>
+
+<details open>
+<summary>MacOS</summary>
+
+Use a package manager like [Homebrew](https://brew.sh) to get the necessary dependencies.
+
+````bash
+$ brew install cmake openblas lapack armadillo
```
Comment (zoq) [inline README.md]: See comment above.

File: README.md [hunk 4]
```diff
@@ -15,6 +15,96 @@ gradient-free optimizers, and constrained optimization.
...
+````
+</details>
+
+<details open>
+<summary>Windows</summary>
+
+You can install **ensmallen** directly by using [vcpkg](https://github.com/microsoft/vcpkg).
```
Comment (zoq) [inline README.md]: Wow, we have a windows package, nice.

**Comment by say4n** [comment]
> Looks like we haven't put that information in the README.md file, so maybe we should remove that line and just update the file with the information you put together.

Okay, I'm migrating the changes from README.md to the CONTRIBUTING.md.

File: README.md (hunk 1 above)
Comment (say4n) [reply README.md]: Okay, will remove these.

File: README.md (hunk 2 above)
Comment (say4n) [reply README.md]: @zoq, should I keep g++ and cmake installation as a part of the dependencies?
I think it should be removed (as it expected that the user has these installed).

File: README.md (hunk 3 above)
Comment (say4n) [reply README.md]: Yes, I am removing BLAS.

File: README.md (hunk 2 above)
Comment (zoq) [reply README.md]: Agrred, I think we can use: http://mlpack.org/doc/mlpack-3.2.2/doxygen/build.html as orientation.

Comment (say4n) [reply README.md]: The simple build instructions on the page don't install any deps, however, the detailed build instructions do. We, presently, in **ensmallen**, have the dependency installation as a separate subsection. Will I remove it?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
 uncommenting the code that sets a random seed in `tests/main.cpp` and running
 your test many times.
 
-Information on how to build and run the tests is in the main README.md file.
+### Install Dependencies
+
+<details open>
```
Comment (zoq) [inline CONTRIBUTING.md]: Nice idea.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+<summary>Linux</summary>
+
+Use your distributions' package manager to install the required dependencies. The commands shown below are for Ubuntu.
```
Comment (zoq) [inline CONTRIBUTING.md]: for Ubuntu and Debian.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+
+```bash
+$ sudo apt-get update
+$ sudo apt-get install liblapack-dev
```
Comment (zoq) [inline CONTRIBUTING.md]: Already a dependency for the `libarmadillo-dev` package, so we can remove the package here as well.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+```
+</details>
+
+<details open>
+<summary>MacOS</summary>
+
+Use a package manager like [Homebrew](https://brew.sh) to get the necessary dependencies.
+
+````bash
+$ brew install lapack armadillo
```
Comment (zoq) [inline CONTRIBUTING.md]: See comment above.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+You can install **ensmallen** directly by using [vcpkg](https://github.com/microsoft/vcpkg).
+
+```
+vcpkg install ensmallen:x64-windows
+```
+</details>
+
+### Build and Test
+
+This section describes how to build and test the **ensmallen** library from source. **ensmallen** uses CMake as its build system.
```
Comment (zoq) [inline CONTRIBUTING.md]: Maybe we should rephrase it here, as we aren't building ensmallen, it's header only. "This section describes how to build the **ensmallen** tests from source."

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+
+```bash
+$ git clone https://github.com/mlpack/ensmallen
+$ cd ensmallen
+
+# - or -
+
+$ wget http://ensmallen.org/files/ensmallen-latest.tar.gz
+$ tar -xvzpf ensmallen-latest.tar.gz
+$ cd ensmallen-latest
```
Comment (zoq) [inline CONTRIBUTING.md]: WE packaged the latest version, so right now the folder would be `ensmallen-2.10.5` and not `ensmallen-latest`, so I think we can use `ensmallen-x.x.x`.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+$ cd build
+```
+
+Then, run the cmake command followed by the make command in the build directory. If the cmake command fails, you probably have missing dependencies.
+
+```bash
+$ cmake ..
+# or with -w flag to inhibit all warning messages
+# $ cmake -DCMAKE_CXX_FLAGS="-w" -DCMAKE_C_FLAGS="-w" .. #
+$ make -j4
```
Comment (zoq) [inline CONTRIBUTING.md]: Should we use just make?, I guess not everyone has that many cores.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+```
+
+Now, you can either run all of the test or an individual test case with:
```
Comment (zoq) [inline CONTRIBUTING.md]: "[...] all of the tests or [...]"

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+
+```bash
+$ ./ensmallen_tests
+$ ./ensmallen_tests <test name>
+```
+
+Ensmallen uses [Catch2](https://github.com/catchorg/Catch2) as the unit test framework.
```
Comment (zoq) [inline CONTRIBUTING.md]: Maybe that should come before the test run section.

File: CONTRIBUTING.md (hunk 11 above)
Comment (abernauer) [reply CONTRIBUTING.md]: Yeah agree maybe add an optional line in a comment with building with -j.

File: CONTRIBUTING.md (hunk 9 above)
Comment (abernauer) [reply CONTRIBUTING.md]: Yeah good distinction it's a header-only library don't want to cause confusion in the documentation.

File: CONTRIBUTING.md (hunk 13 above)
Comment (abernauer) [reply CONTRIBUTING.md]: Including this line around the introductory information on line 81 or around there might be preferable.

File: CONTRIBUTING.md (hunk 8 above)
Comment (abernauer) [reply CONTRIBUTING.md]: Addressing the comments about the dependencies and package managers along with other minor gripes should be adequate changes.

File: CONTRIBUTING.md (hunk 7 above)
Comment (say4n) [reply CONTRIBUTING.md]: Okay.

File: CONTRIBUTING.md (hunk 10 above)
Comment (say4n) [reply CONTRIBUTING.md]: Will I keep the semver as x.x.x, or should we bump it manually for every release?

File: CONTRIBUTING.md (hunk 11 above)
Comment (say4n) [reply CONTRIBUTING.md]: I am changing the `make` with parallel jobs as a separate comment in the instructions and defaulting to just `make`.

File: CONTRIBUTING.md (hunk 12 above)
Comment (say4n) [reply CONTRIBUTING.md]: Will I rephrase this?

File: CONTRIBUTING.md (hunk 13 above)
Comment (say4n) [reply CONTRIBUTING.md]: Okay.

File: CONTRIBUTING.md (hunk 10 above)
Comment (zoq) [reply CONTRIBUTING.md]: We could use the actual version here, we just have too update the release script: https://github.com/mlpack/ensmallen/blob/master/rel/ensmallen-release.sh accordingly.

Comment (say4n) [reply CONTRIBUTING.md]: Okay, on it!

Comment (zoq) [reply CONTRIBUTING.md]: Great, thanks!

**Comment by say4n** [comment]
@zoq I have updated the release script to auto update CONTRIBUTING.md on each release.

Also, regarding the CI build failure, looks like it is a random build failure?

**Comment by zoq** [comment]
Yes, it's unrelated, we might have to adjust the tolerance of the test case, but not as part of the PR.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+### Install Dependencies
+
+<details open>
+<summary>Linux</summary>
+
+Use your distributions' package manager to install the required dependencies. The commands shown below are for Ubuntu/Debian.
+
+```bash
+$ sudo apt-get update
+$ sudo apt-get install libarmadillo-dev
```
Comment (rcurtin) [inline CONTRIBUTING.md]: `cmake` is also a dependency, so that may be worth adding here.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
 uncommenting the code that sets a random seed in `tests/main.cpp` and running
 your test many times.
 
-Information on how to build and run the tests is in the main README.md file.
+### Install Dependencies
+
+<details open>
+<summary>Linux</summary>
+
+Use your distributions' package manager to install the required dependencies. The commands shown below are for Ubuntu/Debian.
```
Comment (rcurtin) [inline CONTRIBUTING.md]: If you want, you can also add Fedora/RHEL/CentOS package manager instructions:

```
yum install armadillo-devel cmake
```

And I think in both of these systems, ensmallen can be installed directly, via `libensmallen-dev` on Debian/Ubuntu and `ensmallen-devel` on Fedora/RHEL/CentOS.  I am not sure of the state of packaging on Arch.

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+</details>
+
+<details open>
+<summary>MacOS</summary>
+
+Use a package manager like [Homebrew](https://brew.sh) to get the necessary dependencies.
+
+````bash
+$ brew install armadillo
+````
```
Comment (rcurtin) [inline CONTRIBUTING.md]: Looks like there might be one too many `'s here.  (Not sure if that makes a difference.)

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+$ sudo apt-get install libarmadillo-dev
+```
+</details>
+
+<details open>
+<summary>MacOS</summary>
+
+Use a package manager like [Homebrew](https://brew.sh) to get the necessary dependencies.
+
+````bash
```
Comment (rcurtin) [inline CONTRIBUTING.md]:
```suggestion
```bash
```

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -42,7 +42,95 @@ you have written a test like this, make sure it does not fail often by
...
+```bash
+$ sudo apt-get update
+$ sudo apt-get install libarmadillo-dev
+```
+</details>
+
+<details open>
+<summary>MacOS</summary>
+
+Use a package manager like [Homebrew](https://brew.sh) to get the necessary dependencies.
```
Comment (rcurtin) [inline CONTRIBUTING.md]: Actually it's now even possible to use Homebrew to install ensmallen directly: https://github.com/Homebrew/homebrew-core/pull/48033

I haven't tried it personally though.

File: rel/ensmallen-release.sh [hunk 19]
```diff
@@ -39,6 +39,9 @@ if [ "$#" -eq "4" ]; then
   sed -i 's/ENS_VERSION_NAME[ ]*\".*\"$/ENS_VERSION_NAME \"'"$4"'\"/' include/ensmallen_bits/ens_version.hpp;
 fi
 
+# update CONTRIBUTING.md
+sed -i "s/ensmallen-[0-9]*\.[0-9]*\.[0-9]*/ensmallen-$MAJOR.$MINOR.$PATCH/g" CONTRIBUTING.md;
+
 git pull
 git add include/ensmallen_bits/ens_version.hpp;
```
Comment (rcurtin) [inline rel/ensmallen-release.sh]: Don't forget to add `CONTRIBUTING.md` here also. :+1:

Comment (say4n) [reply rel/ensmallen-release.sh]: Ok, adding it :)

File: CONTRIBUTING.md (hunk 18 above)
Comment (say4n) [reply CONTRIBUTING.md]: Wow, that's very recent! I'll verify and add it then.

File: CONTRIBUTING.md (hunk 17 above)
Comment (say4n) [reply CONTRIBUTING.md]: Removing the extra backticks.

File: CONTRIBUTING.md (hunk 16 above)
Comment (say4n) [reply CONTRIBUTING.md]: Removing it.

File: CONTRIBUTING.md (hunk 14 above)
Comment (say4n) [reply CONTRIBUTING.md]: Ok, I am adding it to both the Linux and macOS steps then.

File: CONTRIBUTING.md (hunk 18 above)
Comment (say4n) [reply CONTRIBUTING.md]: It works, I am adding it.

File: CONTRIBUTING.md (hunk 15 above)
Comment (say4n) [reply CONTRIBUTING.md]: For Arch, I found [this](https://aur.archlinux.org/packages/ensmallen) package.

Comment (rcurtin) [reply CONTRIBUTING.md]: Ah, right, Roberto packaged that.  If you want you can add the Arch command to install the dependencies or the library, but it's no huge deal---up to you.

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -42,7 +42,109 @@ you have written a test like this, make sure it does not fail often by
...
+<details open>
+<summary>MacOS</summary>
+
+You can install **ensmallen** directyl using [Homebrew](https://brew.sh).
+
+```bash
+$ brew install ensmallen
+```
+
+If you want to build **ensmallen** from source, use a package manager like [Homebrew](https://brew.sh) to get the necessary dependencies.
```
[the remark this answers is not in the document]
Comment (say4n) [reply CONTRIBUTING.md]: @rcurtin I have added instructions for building the tests separately.
`brew install ensmallen` does install the dependencies too.

File: CONTRIBUTING.md (hunk 15 above)
Comment (say4n) [reply CONTRIBUTING.md]: Sure, adding it too.

**Comment by say4n** [comment]
Looking forward to contributing more! :)

**Pull request merged**